### PR TITLE
Introduce template form of skip to represent deprecated variables

### PR DESCRIPTION
**Pull request by ephphatha**
Also ensuring that size_t is used consistently for size/offset types.

### CODE REVIEW

File: Source/loadsave.cpp [hunk 1]
```diff
@@ -1666,10 +1700,10 @@ static void SaveMonster(SaveHelper *file, int i)
 	file->Skip(1); // Alignment
 	file->WriteLE<uint16_t>(pMonster->mExp);
 
-	file->WriteLE<uint8_t>(pMonster->mHit < UINT8_MAX ? pMonster->mHit : UINT8_MAX); // For backwards compatibility
+	file->WriteLE<uint8_t>(std::min<uint16_t>(pMonster->mHit, std::numeric_limits<uint8_t>::max())); // For backwards compatibility
 	file->WriteLE<uint8_t>(pMonster->mMinDamage);
 	file->WriteLE<uint8_t>(pMonster->mMaxDamage);
-	file->WriteLE<uint8_t>(pMonster->mHit2 < UINT8_MAX ? pMonster->mHit2 : UINT8_MAX); // For backwards compatibility
+	file->WriteLE<uint8_t>(std::min<uint16_t>(pMonster->mHit2, std::numeric_limits<uint8_t>::max())); // For backwards compatibility
```
Comment (ephphatha) [inline Source/loadsave.cpp:1706]: Not happy with this but it avoids comparisons between different types and the downcast is recognised as non-truncating in msvc at least.

File: Source/loadsave.cpp [hunk 2]
```diff
@@ -1285,8 +1297,10 @@ static void SaveItem(SaveHelper *file, ItemStruct *pItem)
 	}
 
 	file->WriteLE<int32_t>(pItem->_iSeed);
-	file->WriteLE<int16_t>(pItem->_iCreateInfo);
-	file->Skip(2); // Alignment
+
+	file->WriteLE<uint16_t>(pItem->_iCreateInfo);
+	file->Skip(2); // Alignment, _iCreateInfo was written as a 4 byte LE value but only requires 2 bytes
```
Comment (AJenbo) [inline Source/loadsave.cpp:1302]: I don't think this is why alignment happen for structs https://en.m.wikipedia.org/wiki/Data_structure_alignment

Maybe you are not aware but the original save files would mostly just memcpy structs to the file buffer

Comment (ephphatha) [reply Source/loadsave.cpp:1302]: This was a bit of a guess based on the comments in devilution for certain structs:
https://github.com/diasurgical/devilution/blob/e426c38adb5df05b8e2c02a607529bdc913c281b/structs.h#L241-L247

Situations like this looked more like type conversion/aliasing? compared to other instances where a group of short/char variables needed padding to get back to 4 byte alignment.

Comment (AJenbo) [reply Source/loadsave.cpp:1302]: I think that comment is wrong, the compiler never was able to do this when we worked with it and this looks more like just regular padding.

Comment (ephphatha) [reply Source/loadsave.cpp:1302]: Ahh, in that case I'll revert those comment changes. The other comments I updated with `packed _var, _var, ...` were also more helpers for myself along the same lines so I'll revert those as well.

File: Source/loadsave.cpp [hunk 3]
```diff
@@ -1497,17 +1510,16 @@ static void SavePlayer(SaveHelper *file, int p)
 	file->WriteLE<int32_t>(player._pVar5);
 	file->WriteLE<int32_t>(player.position.offset2.deltaX);
 	file->WriteLE<int32_t>(player.position.offset2.deltaY);
-	// Write actionFrame for vanilla compatibility
-	file->WriteLE<int32_t>(0);
+	file->WriteLE<int32_t>(0); // Write actionFrame for vanilla compatibility
```
Comment (AJenbo) [inline Source/loadsave.cpp:1513]:
```suggestion
	file->Skip<int32_t>(); // Skip _pVar5
```

File: Source/loadsave.cpp [hunk 4]
```diff
@@ -1497,17 +1510,16 @@ static void SavePlayer(SaveHelper *file, int p)
...
 	for (uint8_t i = 0; i < giNumberOfLevels; i++)
 		file->WriteLE<uint8_t>(player._pLvlVisited[i] ? 1 : 0);
 	for (uint8_t i = 0; i < giNumberOfLevels; i++)
 		file->WriteLE<uint8_t>(player._pSLvlVisited[i] ? 1 : 0); // only 10 used
 
 	file->Skip(2); // Alignment
 
-	// Write _pGFXLoad for vanilla compatibility
-	file->WriteLE<int32_t>(0);
+	file->WriteLE<int32_t>(0); // Write _pGFXLoad for vanilla compatibility
```
Comment (AJenbo) [inline Source/loadsave.cpp:1522]:
```suggestion
	file->Skip<int32_t>(); // Skip _pGFXLoad
```

File: Source/loadsave.cpp [hunk 5]
```diff
@@ -1639,8 +1651,7 @@ static void SaveMonster(SaveHelper *file, int i)
 	file->WriteLE<int32_t>(pMonster->position.temp.y);
 	file->WriteLE<int32_t>(pMonster->position.offset2.deltaX);
 	file->WriteLE<int32_t>(pMonster->position.offset2.deltaY);
-	// Write actionFrame for vanilla compatibility
-	file->WriteLE<int32_t>(0);
+	file->WriteLE<int32_t>(0); // Write actionFrame for vanilla compatibility
```
Comment (AJenbo) [inline Source/loadsave.cpp:1654]:
```suggestion
	file->Skip<int32_t>(); // Skip _mVar8
```

File: Source/loadsave.cpp [hunk 6]
```diff
@@ -150,13 +156,19 @@ class SaveHelper {
...
-	bool IsValid(uint32_t len = 1)
+	bool IsValid(size_t len = 1)
 	{
 		return m_buffer_ != nullptr
 		    && m_capacity_ >= (m_cur_ + len);
 	}
 
-	void Skip(uint32_t len)
+	template<typename T>
+	constexpr void Skip(void)
```
Comment (glebm) [inline Source/loadsave.cpp:166]: The `void` is unnecessary in C++

```suggestion
	constexpr void Skip()
```

(same above in LoadHelper)

File: Source/loadsave.cpp [hunk 7]
```diff
@@ -1497,17 +1510,16 @@ static void SavePlayer(SaveHelper *file, int p)
 	file->WriteLE<int32_t>(player._pVar5);
 	file->WriteLE<int32_t>(player.position.offset2.deltaX);
 	file->WriteLE<int32_t>(player.position.offset2.deltaY);
-	// Write actionFrame for vanilla compatibility
```
Comment (qndel) [inline Source/loadsave.cpp:1500]: what happened to this?

File: Source/loadsave.cpp [hunk 8]
```diff
@@ -1298,7 +1310,8 @@ static void SaveItem(SaveHelper *file, ItemStruct *pItem)
 	file->WriteLE<int32_t>(ItemAnimWidth);
 	// write _iAnimWidth2 for vanilla compatibility
 	file->WriteLE<int32_t>(CalculateWidth2(ItemAnimWidth));
-	file->Skip(4); // Unused since 1.02
+	file->Skip<uint32_t>(); // _delFlag, unused since 1.02
+
```
Comment (AJenbo) [inline Source/loadsave.cpp:1314]:
```suggestion
```

File: Source/loadsave.cpp [hunk 9]
```diff
@@ -1497,17 +1510,16 @@ static void SavePlayer(SaveHelper *file, int p)
 	file->WriteLE<int32_t>(player._pVar5);
 	file->WriteLE<int32_t>(player.position.offset2.deltaX);
 	file->WriteLE<int32_t>(player.position.offset2.deltaY);
-	// Write actionFrame for vanilla compatibility
-	file->WriteLE<int32_t>(0);
+	file->Skip<int32_t>(); // Skip _pVar8
+
```
Comment (AJenbo) [inline Source/loadsave.cpp:1513]:
```suggestion
```

**Review by AJenbo — APPROVED** [review]

**Pull request merged**